Add WebProcess initial implementation of background fetch

youennf opened the pull request:
#### 20d85cda31cf7607c143a7a0f9608fa794f1fcf4
<pre>
Add WebProcess initial implementation of background fetch
<a href="https://bugs.webkit.org/show_bug.cgi?id=252036">https://bugs.webkit.org/show_bug.cgi?id=252036</a>
rdar://problem/105262330

Reviewed by Chris Dumez.

Implement BackgroundFetchManager and BackgroundFetchRegistration by using SWClientConnection new functions.
Implement IPC plmubing up to SWServer.

* LayoutTests/imported/w3c/web-platform-tests/background-fetch/abort.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/content-security-policy.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/fetch-uploads.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/fetch.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/get-ids.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/get.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/match.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/mixed-content-and-allowed-schemes.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/port-blocking.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/update-ui.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fetch/private-network-access/service-worker-background-fetch.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/resources/testdriver-vendor.js:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/workers/service/SWClientConnection.cpp:
(WebCore::SWClientConnection::fromScriptExecutionContext):
(WebCore::SWClientConnection::updateBackgroundFetchRegistration):
* Source/WebCore/workers/service/SWClientConnection.h:
* Source/WebCore/workers/service/WorkerSWClientConnection.cpp:
(WebCore::WorkerSWClientConnection::~WorkerSWClientConnection):
(WebCore::WorkerSWClientConnection::subscribeToPushService):
(WebCore::WorkerSWClientConnection::unsubscribeFromPushService):
(WebCore::WorkerSWClientConnection::getPushSubscription):
(WebCore::WorkerSWClientConnection::getPushPermissionState):
(WebCore::WorkerSWClientConnection::getNotifications):
(WebCore::WorkerSWClientConnection::startBackgroundFetch):
(WebCore::WorkerSWClientConnection::backgroundFetchInformation):
(WebCore::WorkerSWClientConnection::backgroundFetchIdentifiers):
(WebCore::WorkerSWClientConnection::abortBackgroundFetch):
(WebCore::WorkerSWClientConnection::matchBackgroundFetch):
(WebCore::toCrossThreadData):
(WebCore::fromCrossThreadData):
(WebCore::WorkerSWClientConnection::retrieveRecordResponse):
(WebCore::WorkerSWClientConnection::retrieveRecordResponseBody):
* Source/WebCore/workers/service/WorkerSWClientConnection.h:
* Source/WebCore/workers/service/background-fetch/BackgroundFetchManager.cpp:
(WebCore::BackgroundFetchManager::BackgroundFetchManager):
(WebCore::buildBackgroundFetchRequests):
(WebCore::BackgroundFetchManager::backgroundFetchRegistrationInstance):
(WebCore::BackgroundFetchManager::fetch):
(WebCore::BackgroundFetchManager::get):
(WebCore::BackgroundFetchManager::getIds):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchManager.h:
(WebCore::BackgroundFetchManager::existingBackgroundFetchRegistration):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchRecordInformation.h:
* Source/WebCore/workers/service/background-fetch/BackgroundFetchRegistration.cpp:
(WebCore::BackgroundFetchRegistration::updateIfExisting):
(WebCore::BackgroundFetchRegistration::abort):
(WebCore::requestFromInfo):
(WebCore::createRecord):
(WebCore::BackgroundFetchRegistration::match):
(WebCore::BackgroundFetchRegistration::matchAll):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchRegistration.h:
* Source/WebCore/workers/service/server/SWServer.cpp:
(WebCore::SWServer::Connection::startBackgroundFetch):
(WebCore::SWServer::Connection::backgroundFetchInformation):
(WebCore::SWServer::Connection::backgroundFetchIdentifiers):
(WebCore::SWServer::Connection::abortBackgroundFetch):
(WebCore::SWServer::Connection::matchBackgroundFetch):
(WebCore::SWServer::Connection::retrieveRecordResponse):
(WebCore::SWServer::Connection::retrieveRecordResponseBody):
* Source/WebCore/workers/service/server/SWServer.h:
* Source/WebCore/workers/service/server/SWServerToContextConnection.h:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.cpp:
(WebKit::WebSWServerConnection::WebSWServerConnection):
(WebKit::WebSWServerConnection::controlClient):
(WebKit::WebSWServerConnection::createFetchTask):
(WebKit::WebSWServerConnection::startFetch):
(WebKit::WebSWServerConnection::postMessageToServiceWorker):
(WebKit::WebSWServerConnection::scheduleJobInServer):
(WebKit::WebSWServerConnection::clientURLFromIdentifier):
(WebKit::WebSWServerConnection::scheduleUnregisterJobInServer):
(WebKit::WebSWServerConnection::postMessageToServiceWorkerClient):
(WebKit::WebSWServerConnection::registerServiceWorkerClient):
(WebKit::WebSWServerConnection::unregisterServiceWorkerClient):
(WebKit::WebSWServerConnection::updateThrottleState):
(WebKit::WebSWServerConnection::subscribeToPushService):
(WebKit::WebSWServerConnection::unsubscribeFromPushService):
(WebKit::WebSWServerConnection::getPushSubscription):
(WebKit::WebSWServerConnection::getPushPermissionState):
(WebKit::WebSWServerConnection::contextConnectionCreated):
(WebKit::WebSWServerConnection::fetchTaskTimedOut):
(WebKit::WebSWServerConnection::gatherClientData):
(WebKit::WebSWServerConnection::updateBackgroundFetchRegistration):
(WebKit::WebSWServerConnection::retrieveRecordResponseBody):
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.h:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.messages.in:
* Source/WebKit/Scripts/webkit/messages.py:
(serialized_identifiers):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Storage/RetrieveRecordResponseBodyCallbackIdentifier.h: Added.
* Source/WebKit/WebProcess/Storage/WebSWClientConnection.cpp:
(WebKit::callExceptionOrResultCallback):
(WebKit::WebSWClientConnection::subscribeToPushService):
(WebKit::WebSWClientConnection::unsubscribeFromPushService):
(WebKit::WebSWClientConnection::getPushSubscription):
(WebKit::WebSWClientConnection::getNavigationPreloadState):
(WebKit::WebSWClientConnection::startBackgroundFetch):
(WebKit::WebSWClientConnection::backgroundFetchInformation):
(WebKit::WebSWClientConnection::backgroundFetchIdentifiers):
(WebKit::WebSWClientConnection::abortBackgroundFetch):
(WebKit::WebSWClientConnection::matchBackgroundFetch):
(WebKit::WebSWClientConnection::retrieveRecordResponse):
(WebKit::WebSWClientConnection::retrieveRecordResponseBody):
(WebKit::WebSWClientConnection::notifyRecordResponseBodyChunk):
(WebKit::WebSWClientConnection::notifyRecordResponseBodyEnd):
* Source/WebKit/WebProcess/Storage/WebSWClientConnection.h:
* Source/WebKit/WebProcess/Storage/WebSWClientConnection.messages.in:

Canonical link: <a href="https://commits.webkit.org/260602@main">https://commits.webkit.org/260602@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f3fca1db84ac8c4df869cf562bb6fd0d65433936

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108853 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17956 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41689 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/390 "Failed to checkout and rebase branch from PR 9913") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/118185 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/112735 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/19414 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9235 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101090 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114623 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/14568 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/97771 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/42605 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [❌ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/112378 "Found 1 webkitpy python3 test failure: webkit.messages_unittest.GeneratedFileContentsTest.test_message_argument_description") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/96540 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/29414 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/84391 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10731 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/30761 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11486 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/7694 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16875 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/50362 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7323 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13073 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/workers/service/background-fetch/BackgroundFetchManager.cpp:
```diff
@@ -28,31 +28,147 @@
+static ExceptionOr<Vector<Ref<FetchRequest>>> buildBackgroundFetchRequests(ScriptExecutionContext& context, BackgroundFetchManager::Requests&& backgroundFetchRequests)
 {
-    promise.reject(Exception { NotSupportedError });
+    if (std::holds_alternative<RefPtr<FetchRequest>>(backgroundFetchRequests)) {
```
Any reason we're not using WTF::switchOn() instead of std::holds_alternative + std::get?

youennf replied:
I'll switch to switch ;)

youennf replied:
Done

cdumez commented on Source/WebCore/workers/service/background-fetch/BackgroundFetchManager.cpp:
```diff
@@ -28,31 +28,147 @@
+
+    auto registration = result.iterator->value;
+    if (!result.isNewEntry)
+        registration->updateInformation(data);
```
use-after-move of `data`?

youennf replied:
Not really, since the lambda is not called if there is already an entry. I'll think of a better way to write it.

cdumez replied:
May look clearer to capture `&data` instead.

cdumez replied:
or simply `[&]`

cdumez commented on Source/WebCore/workers/service/background-fetch/BackgroundFetchRegistration.cpp:
```diff
@@ -41,6 +45,15 @@ namespace WebCore {
+    RefPtr registration = container ? container->registration(information.registrationIdentifier) : nullptr;
+    RefPtr manager = registration ? ServiceWorkerRegistrationBackgroundFetchAPI::backgroundFetchIfCreated(*registration) : nullptr;
+    if (auto backgroundFetrchRegistration = manager ? manager->existingBackgroundFetchRegistration(information.identifier) : nullptr)
+        backgroundFetrchRegistration->updateInformation(information);
```
typo: backgroundFetrchRegistration

cdumez commented on Source/WebKit/WebProcess/Storage/WebSWClientConnection.cpp:
```diff
@@ -321,12 +327,74 @@ void WebSWClientConnection::setNavigationPreloadHeaderValue(WebCore::ServiceWork
+
+void WebSWClientConnection::notifyRecordResponseBodyEnd(RetrieveRecordResponseBodyCallbackIdentifier identifier, WebCore::ResourceError&& error)
+{
+    auto iterator = m_retrieveRecordResponseBodyCallbacks.find(identifier);
```
Maybe we can call take() here since we know we'll remove from the map?

cdumez reviewed (approved):
LGTM

youennf committed:
Add WebProcess initial implementation of background fetch
https://bugs.webkit.org/show_bug.cgi?id=252036
rdar://problem/105262330

Reviewed by Chris Dumez.

Implement BackgroundFetchManager and BackgroundFetchRegistration by using SWClientConnection new functions.
Implement IPC plmubing up to SWServer.

* LayoutTests/imported/w3c/web-platform-tests/background-fetch/abort.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/content-security-policy.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/fetch-uploads.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/fetch.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/get-ids.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/get.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/match.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/mixed-content-and-allowed-schemes.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/port-blocking.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/update-ui.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fetch/private-network-access/service-worker-background-fetch.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/resources/testdriver-vendor.js:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/workers/service/SWClientConnection.cpp:
(WebCore::SWClientConnection::fromScriptExecutionContext):
(WebCore::SWClientConnection::updateBackgroundFetchRegistration):
* Source/WebCore/workers/service/SWClientConnection.h:
* Source/WebCore/workers/service/WorkerSWClientConnection.cpp:
(WebCore::WorkerSWClientConnection::~WorkerSWClientConnection):
(WebCore::WorkerSWClientConnection::subscribeToPushService):
(WebCore::WorkerSWClientConnection::unsubscribeFromPushService):
(WebCore::WorkerSWClientConnection::getPushSubscription):
(WebCore::WorkerSWClientConnection::getPushPermissionState):
(WebCore::WorkerSWClientConnection::getNotifications):
(WebCore::WorkerSWClientConnection::startBackgroundFetch):
(WebCore::WorkerSWClientConnection::backgroundFetchInformation):
(WebCore::WorkerSWClientConnection::backgroundFetchIdentifiers):
(WebCore::WorkerSWClientConnection::abortBackgroundFetch):
(WebCore::WorkerSWClientConnection::matchBackgroundFetch):
(WebCore::toCrossThreadData):
(WebCore::fromCrossThreadData):
(WebCore::WorkerSWClientConnection::retrieveRecordResponse):
(WebCore::WorkerSWClientConnection::retrieveRecordResponseBody):
* Source/WebCore/workers/service/WorkerSWClientConnection.h:
* Source/WebCore/workers/service/background-fetch/BackgroundFetchManager.cpp:
(WebCore::BackgroundFetchManager::BackgroundFetchManager):
(WebCore::buildBackgroundFetchRequests):
(WebCore::BackgroundFetchManager::backgroundFetchRegistrationInstance):
(WebCore::BackgroundFetchManager::fetch):
(WebCore::BackgroundFetchManager::get):
(WebCore::BackgroundFetchManager::getIds):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchManager.h:
(WebCore::BackgroundFetchManager::existingBackgroundFetchRegistration):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchRecordInformation.h:
* Source/WebCore/workers/service/background-fetch/BackgroundFetchRegistration.cpp:
(WebCore::BackgroundFetchRegistration::updateIfExisting):
(WebCore::BackgroundFetchRegistration::abort):
(WebCore::requestFromInfo):
(WebCore::createRecord):
(WebCore::BackgroundFetchRegistration::match):
(WebCore::BackgroundFetchRegistration::matchAll):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchRegistration.h:
* Source/WebCore/workers/service/server/SWServer.cpp:
(WebCore::SWServer::Connection::startBackgroundFetch):
(WebCore::SWServer::Connection::backgroundFetchInformation):
(WebCore::SWServer::Connection::backgroundFetchIdentifiers):
(WebCore::SWServer::Connection::abortBackgroundFetch):
(WebCore::SWServer::Connection::matchBackgroundFetch):
(WebCore::SWServer::Connection::retrieveRecordResponse):
(WebCore::SWServer::Connection::retrieveRecordResponseBody):
* Source/WebCore/workers/service/server/SWServer.h:
* Source/WebCore/workers/service/server/SWServerToContextConnection.h:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.cpp:
(WebKit::WebSWServerConnection::WebSWServerConnection):
(WebKit::WebSWServerConnection::controlClient):
(WebKit::WebSWServerConnection::createFetchTask):
(WebKit::WebSWServerConnection::startFetch):
(WebKit::WebSWServerConnection::postMessageToServiceWorker):
(WebKit::WebSWServerConnection::scheduleJobInServer):
(WebKit::WebSWServerConnection::clientURLFromIdentifier):
(WebKit::WebSWServerConnection::scheduleUnregisterJobInServer):
(WebKit::WebSWServerConnection::postMessageToServiceWorkerClient):
(WebKit::WebSWServerConnection::registerServiceWorkerClient):
(WebKit::WebSWServerConnection::unregisterServiceWorkerClient):
(WebKit::WebSWServerConnection::updateThrottleState):
(WebKit::WebSWServerConnection::subscribeToPushService):
(WebKit::WebSWServerConnection::unsubscribeFromPushService):
(WebKit::WebSWServerConnection::getPushSubscription):
(WebKit::WebSWServerConnection::getPushPermissionState):
(WebKit::WebSWServerConnection::contextConnectionCreated):
(WebKit::WebSWServerConnection::fetchTaskTimedOut):
(WebKit::WebSWServerConnection::gatherClientData):
(WebKit::WebSWServerConnection::updateBackgroundFetchRegistration):
(WebKit::WebSWServerConnection::retrieveRecordResponseBody):
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.h:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.messages.in:
* Source/WebKit/Scripts/webkit/messages.py:
(serialized_identifiers):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Storage/RetrieveRecordResponseBodyCallbackIdentifier.h: Added.
* Source/WebKit/WebProcess/Storage/WebSWClientConnection.cpp:
(WebKit::callExceptionOrResultCallback):
(WebKit::WebSWClientConnection::subscribeToPushService):
(WebKit::WebSWClientConnection::unsubscribeFromPushService):
(WebKit::WebSWClientConnection::getPushSubscription):
(WebKit::WebSWClientConnection::getNavigationPreloadState):
(WebKit::WebSWClientConnection::startBackgroundFetch):
(WebKit::WebSWClientConnection::backgroundFetchInformation):
(WebKit::WebSWClientConnection::backgroundFetchIdentifiers):
(WebKit::WebSWClientConnection::abortBackgroundFetch):
(WebKit::WebSWClientConnection::matchBackgroundFetch):
(WebKit::WebSWClientConnection::retrieveRecordResponse):
(WebKit::WebSWClientConnection::retrieveRecordResponseBody):
(WebKit::WebSWClientConnection::notifyRecordResponseBodyChunk):
(WebKit::WebSWClientConnection::notifyRecordResponseBodyEnd):
* Source/WebKit/WebProcess/Storage/WebSWClientConnection.h:
* Source/WebKit/WebProcess/Storage/WebSWClientConnection.messages.in:

Canonical link: https://commits.webkit.org/260602@main
